File: poly_apps/nuxt_main/scripts/start.js
#!/usr/bin/env node
/**
 * Nuxt Main Start Script - Full Featured Launcher (JavaScript Version)
 *
 * Equivalent to start.ps1 with all features
 *
 * Usage:
 *   node start.js                    # Interactive menu
 *   node start.js <app>              # Direct launch in debug mode
 *   node start.js <app> debug        # Direct launch in debug mode
 *   node start.js <app> build        # Direct launch in build mode
 *   node start.js help               # Show help
 *   node start.js list               # List available apps
 *   node start.js -MultiApps a,b     # Multi-app debug mode
 *
 * Examples:
 *   node start.js pymatrix           # Start pyMatrix in debug mode
 *   node start.js ittools build      # Build ITTools
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const { spawn, execSync } = require('child_process');

// ============================================================
// Global Variables
// ============================================================
const ORIGINAL_WORKING_DIR = process.cwd();
const SCRIPT_DIR = __dirname;
const APP_DIR = path.resolve(SCRIPT_DIR, '..');
const SWITCH_APP_SCRIPT = path.join(SCRIPT_DIR, 'switch-app.js');

const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
  gray: '\x1b[90m'
};

function log(message, color = 'reset') {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

function separator(color = 'cyan') {
  log('='.repeat(79), color);
}

// ============================================================
// App Configuration Scanner
// ============================================================
function scanAppConfigs() {
  const appsDir = path.join(APP_DIR, 'apps');
  const appConfigs = {};

  if (!fs.existsSync(appsDir)) {
    return appConfigs;
  }

  const entries = fs.readdirSync(appsDir, { withFileTypes: true });

  entries.forEach(entry => {
    if (entry.isDirectory() && entry.name.startsWith('app_')) {
      const appName = entry.name.replace(/^app_/, '');
      const configPath = path.join(appsDir, entry.name, 'app-config.json');

      if (fs.existsSync(configPath)) {
        try {
          const configData = JSON.parse(fs.readFileSync(configPath, 'utf8'));
          appConfigs[appName] = {
            name: appName,
            displayName: configData.displayName || configData.display_name || appName,
            port: configData.port || 3000,
            devCommand: configData.devCommand || configData.dev_command || 'nuxt dev',
            buildCommand: configData.buildCommand || configData.build_command || 'nuxt build'
          };
        } catch (e) {
          // Skip invalid config
        }
      }
    }
  });

  return appConfigs;
}

// ============================================================
// Interactive Menu
// ============================================================
async function showInteractiveMenu(appConfigs, savedState) {
  const apps = Object.keys(appConfigs).sort();
  let selectedIndex = savedState.selectedIndex || 0;
  let mode = savedState.mode || 'debug';

  if (selectedIndex >= apps.length) selectedIndex = 0;

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  readline.emitKeypressEvents(process.stdin);
  if (process.stdin.isTTY) {
    process.stdin.setRawMode(true);
  }

  function render() {
    console.clear();
    separator('green');
    log('  NUXT MULTI-APP LAUNCHER - INTERACTIVE MENU', 'green');
    separator('green');
    log('');
    log('Use Arrow Keys to navigate, Enter to select, Q to quit', 'yellow');
    log('Press D for Debug mode, B for Build mode', 'yellow');
    log('');
    log(`Current Mode: ${mode.toUpperCase()}`, mode === 'debug' ? 'cyan' : 'magenta');
    log('');

    apps.forEach((appName, index) => {
      const config = appConfigs[appName];
      const prefix = index === selectedIndex ? '→' : ' ';
      const color = index === selectedIndex ? 'cyan' : 'white';
      log(`  ${prefix} [${index + 1}] ${config.displayName} (${appName}) - Port ${config.port}`, color);
    });

    log('');
    separator('green');
  }

  return new Promise((resolve) => {
    render();

    process.stdin.on('keypress', (str, key) => {
      if (key.name === 'q' || (key.ctrl && key.name === 'c')) {
        process.stdin.setRawMode(false);
        rl.close();
        process.exit(0);
      }

      if (key.name === 'up') {
        selectedIndex = selectedIndex > 0 ? selectedIndex - 1 : apps.length - 1;
        render();
      } else if (key.name === 'down') {
        selectedIndex = selectedIndex < apps.length - 1 ? selectedIndex + 1 : 0;
        render();
      } else if (key.name === 'd') {
        mode = 'debug';
        render();
      } else if (key.name === 'b') {
        mode = 'build';
        render();
      } else if (key.name === 'return') {
        process.stdin.setRawMode(false);
        rl.close();
        const appName = apps[selectedIndex];
        resolve({
          selectedApp: appConfigs[appName],
          mode: mode,
          selectedIndex: selectedIndex
        });
      }
    });
  });
}

// ============================================================
// Menu State Management
// ============================================================
function getMenuState() {
  const statePath = path.join(APP_DIR, '.menu-state.json');
  try {
    if (fs.existsSync(statePath)) {
      return JSON.parse(fs.readFileSync(statePath, 'utf8'));
    }
  } catch (e) {
    // Return default
  }
  return { selectedIndex: 0, mode: 'debug' };
}

function saveMenuState(selectedIndex, mode) {
  const statePath = path.join(APP_DIR, '.menu-state.json');
  fs.writeFileSync(statePath, JSON.stringify({ selectedIndex, mode }), 'utf8');
}

// ============================================================
// Parse Arguments
// ============================================================
function parseArgs() {
  const args = process.argv.slice(2);
  const result = {
    appName: '',
    mode: 'debug',
    multiApps: [],
    showHelp: false,
    showList: false
  };

  if (args.length === 0) {
    return result; // Interactive mode
  }

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];

    if (arg === 'help' || arg === '-h' || arg === '--help') {
      result.showHelp = true;
      return result;
    }

    if (arg === 'list' || arg === '-list' || arg === '--list') {
      result.showList = true;
      return result;
    }

    if (arg === '-MultiApps' || arg === '--multi' || arg === 'multi') {
      const nextArg = args[i + 1];
      if (nextArg) {
        result.multiApps = nextArg.split(',').map(s => s.trim());
        i++;
      }
      continue;
    }

    if (i === 0) {
      result.appName = arg;
    } else if (i === 1) {
      if (arg === 'debug' || arg === 'build') {
        result.mode = arg;
      }
    }
  }

  return result;
}

// ============================================================
// Show Help
// ============================================================
function showHelp(appConfigs) {
  log('');
  separator('green');
  log('  NUXT MAIN LAUNCHER - HELP', 'green');
  separator('green');
  log('');
  log('Usage:', 'yellow');
  log('  node start.js                    Interactive menu', 'white');
  log('  node start.js <app>              Direct launch in debug mode', 'white');
  log('  node start.js <app> debug        Direct launch in debug mode', 'white');
  log('  node start.js <app> build        Direct launch in build mode', 'white');
  log('  node start.js list               List available apps', 'white');
  log('  node start.js help               Show this help', 'white');
  log('  node start.js -MultiApps a,b     Multi-app debug mode', 'white');
  log('');
  log('Available Apps:', 'yellow');
  Object.keys(appConfigs).sort().forEach(key => {
    const config = appConfigs[key];
    log(`  - ${key}`, 'cyan');
    log(`    (${config.displayName} on port ${config.port})`, 'gray');
  });
  log('');
  log('Examples:', 'yellow');
  log('  node start.js pymatrix           Start pyMatrix in debug mode', 'white');
  log('  node start.js ittools build      Build ITTools', 'white');
  log('');
  separator('green');
}

// ============================================================
// Show List
// ============================================================
function showList(appConfigs) {
  log('');
  separator('green');
  log('  AVAILABLE APPLICATIONS', 'green');
  separator('green');
  log('');
  Object.keys(appConfigs).sort().forEach(key => {
    const config = appConfigs[key];
    log('  App Name     : ', 'yellow');
    log(`  ${key}`, 'cyan');
    log('  Display Name : ', 'yellow');
    log(`  ${config.displayName}`, 'white');
    log('  Port         : ', 'yellow');
    log(`  ${config.port}`, 'white');
    log('');
  });
  separator('green');
}

// ============================================================
// Execute Command with Error Handling
// ============================================================
function executeCommand(command, args, description, cwd, env) {
  log('');
  separator('magenta');
  log(`  ${description}`, 'magenta');
  separator('magenta');
  log(`[COMMAND] ${command} ${args.join(' ')}`, 'yellow');
  log('');

  try {
    execSync(`${command} ${args.join(' ')}`, {
      cwd: cwd,
      env: { ...process.env, ...env },
      stdio: 'inherit'
    });
    log('');
    log(`[SUCCESS] ${description} completed`, 'green');
    separator('magenta');
    log('');
    return true;
  } catch (error) {
    log('');
    log(`[ERROR] ${description} failed`, 'red');
    log(`[ERROR] ${error.message}`, 'red');
    separator('magenta');
    log('');
    return false;
  }
}

// ============================================================
// Main Function
// ============================================================
async function main() {
  log('');
  separator('cyan');
  log('  NUXT MAIN APPLICATION LAUNCHER - INITIALIZATION', 'cyan');
  separator('cyan');
  log('');
  log('[TRACE] Script Initialization:', 'yellow');
  log(`  > Original Working Directory: ${ORIGINAL_WORKING_DIR}`, 'white');
  log(`  > Script Directory: ${SCRIPT_DIR}`, 'white');
  log(`  > Application Directory: ${APP_DIR}`, 'white');
  log('');

  // Scan app configs
  log('[TRACE] Scanning application configurations...', 'yellow');
  const appConfigs = scanAppConfigs();
  log(`  [OK] Found ${Object.keys(appConfigs).length} applications`, 'green');
  log('');

  // Change to app directory
  log('[TRACE] Changing Directory to Application Root:', 'yellow');
  log(`  > Set Working Directory: ${APP_DIR}`, 'white');
  process.chdir(APP_DIR);
  log(`  [OK] Current Location: ${process.cwd()}`, 'green');
  log('');
  separator('cyan');
  log('');

  // Parse arguments
  const parsedArgs = parseArgs();

  if (parsedArgs.showHelp) {
    showHelp(appConfigs);
    process.exit(0);
  }

  if (parsedArgs.showList) {
    showList(appConfigs);
    process.exit(0);
  }

  // Determine mode
  let selectedApp;
  let mode;
  let selectedIndex = 0;

  if (parsedArgs.appName) {
    // Command-line mode
    if (!appConfigs[parsedArgs.appName]) {
      log('');
      separator('red');
      log(`  ERROR: Unknown application '${parsedArgs.appName}'`, 'red');
      separator('red');
      log('');
      log('Available apps:', 'yellow');
      Object.keys(appConfigs).sort().forEach(key => {
        log(`  - ${key}`, 'cyan');
      });
      log('');
      process.exit(1);
    }

    selectedApp = appConfigs[parsedArgs.appName];
    mode = parsedArgs.mode;

    log('');
    separator('green');
    log('  COMMAND-LINE MODE ACTIVATED', 'green');
    separator('green');
    log(`  Selected App: ${selectedApp.displayName}`, 'cyan');
    log(`  Mode: ${mode}`, 'cyan');
    separator('green');
    log('');
  } else {
    // Interactive mode
    const savedState = getMenuState();
    const menuResult = await showInteractiveMenu(appConfigs, savedState);
    selectedApp = menuResult.selectedApp;
    mode = menuResult.mode;
    selectedIndex = menuResult.selectedIndex;
    saveMenuState(selectedIndex, mode);
  }

  // Set environment variables
  const env = {
    NUXT_HOST: '0.0.0.0',
    NUXT_PORT: selectedApp.port.toString(),
    APP_ENTRY: selectedApp.name
  };

  log('');
  separator('cyan');
  log('  APPLICATION STARTUP INFO', 'cyan');
  separator('cyan');
  log('');
  log('=== Application Selection ===', 'yellow');
  log(`  Selected App     : ${selectedApp.displayName}`, 'green');
  log(`  Namespace        : ${selectedApp.name}`, 'green');
  log(`  Mode             : ${mode}`, 'green');
  log(`  Port             : ${selectedApp.port}`, 'green');
  log(`  Host             : 0.0.0.0`, 'green');
  log('');
  log('=== Network Configuration ===', 'yellow');
  log(`  Local URL        : http://127.0.0.1:${selectedApp.port}`, 'cyan');
  log(`  Network URL      : http://0.0.0.0:${selectedApp.port}`, 'cyan');
  log('');
  separator('cyan');
  log('');

  // Step 1: Switch pages directory
  const switchSuccess = executeCommand(
    'node',
    [SWITCH_APP_SCRIPT, selectedApp.name],
    'STEP 1: SWITCHING APP PAGES DIRECTORY',
    APP_DIR,
    env
  );

  if (!switchSuccess) {
    log('[ERROR] Failed to switch pages directory', 'red');
    process.exit(1);
  }

  // Step 2: Start application server
  if (mode === 'debug') {
    log('');
    log('[INFO] Opening browser...', 'green');
    const url = `http://127.0.0.1:${selectedApp.port}`;
    try {
      const openCommand = process.platform === 'win32' ? 'start' :
                         process.platform === 'darwin' ? 'open' : 'xdg-open';
      require('child_process').exec(`${openCommand} ${url}`);
    } catch (e) {
      log('[WARNING] Could not open browser automatically', 'yellow');
    }

    executeCommand(
      'node',
      [SWITCH_APP_SCRIPT, selectedApp.name, '--mode', 'dev'],
      'STEP 2: STARTING APPLICATION SERVER (DEBUG MODE)',
      APP_DIR,
      env
    );
  } else {
    executeCommand(
      'node',
      [SWITCH_APP_SCRIPT, selectedApp.name, '--mode', 'build'],
      'STEP 2: BUILDING APPLICATION (BUILD MODE)',
      APP_DIR,
      env
    );
  }

  // Restore original directory
  process.chdir(ORIGINAL_WORKING_DIR);
}

// ============================================================
// Entry Point
// ============================================================
main().catch(error => {
  console.error('[FATAL ERROR]', error);
  process.exit(1);
});

File: poly_apps/nuxt_main/scripts/start.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Nuxt Main Start Script - Full Featured Launcher (Python Version)

Equivalent to start.ps1 with all features

Usage:
    python start.py                    # Interactive menu
    python start.py <app>              # Direct launch in debug mode
    python start.py <app> debug        # Direct launch in debug mode
    python start.py <app> build        # Direct launch in build mode
    python start.py help               # Show help
    python start.py list               # List available apps

Examples:
    python start.py pymatrix           # Start pyMatrix in debug mode
    python start.py ittools build      # Build ITTools
"""

import os
import sys
import json
import subprocess
from pathlib import Path
from typing import Dict, Optional, Tuple

# ============================================================
# Global Variables
# ============================================================
ORIGINAL_WORKING_DIR = os.getcwd()
SCRIPT_DIR = Path(__file__).parent
APP_DIR = SCRIPT_DIR.parent
SWITCH_APP_SCRIPT = SCRIPT_DIR / "switch-app.js"

# ANSI Color Codes
class Colors:
    RESET = '\033[0m'
    BRIGHT = '\033[1m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    WHITE = '\033[37m'
    GRAY = '\033[90m'


def log(message: str, color: str = 'RESET'):
    """Print colored log message"""
    color_code = getattr(Colors, color.upper(), Colors.RESET)
    print(f"{color_code}{message}{Colors.RESET}")


def separator(color: str = 'CYAN'):
    """Print separator line"""
    log('=' * 79, color)


# ============================================================
# App Configuration Scanner
# ============================================================
def scan_app_configs() -> Dict:
    """Scan and load app configurations"""
    apps_dir = APP_DIR / 'apps'
    app_configs = {}

    if not apps_dir.exists():
        return app_configs

    for entry in apps_dir.iterdir():
        if entry.is_dir() and entry.name.startswith('app_'):
            app_name = entry.name.replace('app_', '', 1)
            config_path = entry / 'app-config.json'

            if config_path.exists():
                try:
                    with open(config_path, 'r', encoding='utf-8') as f:
                        config_data = json.load(f)
                        app_configs[app_name] = {
                            'name': app_name,
                            'displayName': config_data.get('displayName', config_data.get('display_name', app_name)),
                            'port': config_data.get('port', 3000),
                            'devCommand': config_data.get('devCommand', config_data.get('dev_command', 'nuxt dev')),
                            'buildCommand': config_data.get('buildCommand', config_data.get('build_command', 'nuxt build'))
                        }
                except Exception:
                    # Skip invalid config
                    pass

    return app_configs


# ============================================================
# Interactive Menu
# ============================================================
def show_interactive_menu(app_configs: Dict, saved_state: Dict) -> Dict:
    """
    Show interactive menu for app selection

    Returns:
        Dict with selectedApp, mode, selectedIndex
    """
    try:
        import readchar
    except ImportError:
        log('[WARNING] readchar not installed, using simplified menu', 'YELLOW')
        return show_simple_menu(app_configs, saved_state)

    apps = sorted(app_configs.keys())
    selected_index = saved_state.get('selectedIndex', 0)
    mode = saved_state.get('mode', 'debug')

    if selected_index >= len(apps):
        selected_index = 0

    def render():
        nonlocal selected_index, mode
        os.system('cls' if os.name == 'nt' else 'clear')
        separator('GREEN')
        log('  NUXT MULTI-APP LAUNCHER - INTERACTIVE MENU', 'GREEN')
        separator('GREEN')
        log('')
        log('Use Arrow Keys to navigate, Enter to select, Q to quit', 'YELLOW')
        log('Press D for Debug mode, B for Build mode', 'YELLOW')
        log('')
        log(f'Current Mode: {mode.upper()}', 'CYAN' if mode == 'debug' else 'MAGENTA')
        log('')

        for index, app_name in enumerate(apps):
            config = app_configs[app_name]
            prefix = '→' if index == selected_index else ' '
            color = 'CYAN' if index == selected_index else 'WHITE'
            log(f'  {prefix} [{index + 1}] {config["displayName"]} ({app_name}) - Port {config["port"]}', color)

        log('')
        separator('GREEN')

    def update_state(new_index=None, new_mode=None):
        nonlocal selected_index, mode
        if new_index is not None:
            selected_index = new_index
        if new_mode is not None:
            mode = new_mode

    while True:
        render()

        try:
            key = readchar.readkey()
        except KeyboardInterrupt:
            sys.exit(0)

        if key == 'q' or key == 'Q':
            sys.exit(0)
        elif key == readchar.key.UP:
            new_index = selected_index - 1 if selected_index > 0 else len(apps) - 1
            update_state(new_index=new_index)
        elif key == readchar.key.DOWN:
            new_index = selected_index + 1 if selected_index < len(apps) - 1 else 0
            update_state(new_index=new_index)
        elif key == 'd' or key == 'D':
            update_state(new_mode='debug')
        elif key == 'b' or key == 'B':
            update_state(new_mode='build')
        elif key == readchar.key.ENTER or key == '\r' or key == '\n':
            app_name = apps[selected_index]
            return {
                'selectedApp': app_configs[app_name],
                'mode': mode,
                'selectedIndex': selected_index
            }


def show_simple_menu(app_configs: Dict, saved_state: Dict) -> Dict:
    """Simplified menu without interactive navigation"""
    apps = sorted(app_configs.keys())

    separator('GREEN')
    log('  NUXT MULTI-APP LAUNCHER - MENU', 'GREEN')
    separator('GREEN')
    log('')

    for index, app_name in enumerate(apps):
        config = app_configs[app_name]
        log(f'  [{index + 1}] {config["displayName"]} ({app_name}) - Port {config["port"]}', 'CYAN')

    log('')
    log('Enter number to select app, or Q to quit:', 'YELLOW')

    try:
        choice = input('> ').strip()
    except KeyboardInterrupt:
        sys.exit(0)

    if choice.lower() == 'q':
        sys.exit(0)

    try:
        index = int(choice) - 1
        if 0 <= index < len(apps):
            app_name = apps[index]
            log('')
            log('Select mode: [1] Debug  [2] Build', 'YELLOW')
            mode_choice = input('> ').strip()
            mode = 'build' if mode_choice == '2' else 'debug'

            return {
                'selectedApp': app_configs[app_name],
                'mode': mode,
                'selectedIndex': index
            }
    except (ValueError, IndexError):
        pass

    log('[ERROR] Invalid selection', 'RED')
    sys.exit(1)


# ============================================================
# Menu State Management
# ============================================================
def get_menu_state() -> Dict:
    """Load saved menu state"""
    state_path = APP_DIR / '.menu-state.json'
    try:
        if state_path.exists():
            with open(state_path, 'r', encoding='utf-8') as f:
                return json.load(f)
    except Exception:
        pass
    return {'selectedIndex': 0, 'mode': 'debug'}


def save_menu_state(selected_index: int, mode: str):
    """Save menu state"""
    state_path = APP_DIR / '.menu-state.json'
    with open(state_path, 'w', encoding='utf-8') as f:
        json.dump({'selectedIndex': selected_index, 'mode': mode}, f)


# ============================================================
# Parse Arguments
# ============================================================
def parse_args() -> Dict:
    """Parse command-line arguments"""
    args = sys.argv[1:]
    result = {
        'appName': '',
        'mode': 'debug',
        'multiApps': [],
        'showHelp': False,
        'showList': False
    }

    if not args:
        return result  # Interactive mode

    i = 0
    while i < len(args):
        arg = args[i]

        if arg in ('help', '-h', '--help'):
            result['showHelp'] = True
            return result

        if arg in ('list', '-list', '--list'):
            result['showList'] = True
            return result

        if arg in ('-MultiApps', '--multi', 'multi'):
            if i + 1 < len(args):
                result['multiApps'] = [s.strip() for s in args[i + 1].split(',')]
                i += 1
            i += 1
            continue

        if i == 0:
            result['appName'] = arg
        elif i == 1:
            if arg in ('debug', 'build'):
                result['mode'] = arg

        i += 1

    return result


# ============================================================
# Show Help
# ============================================================
def show_help(app_configs: Dict):
    """Show help message"""
    log('')
    separator('GREEN')
    log('  NUXT MAIN LAUNCHER - HELP', 'GREEN')
    separator('GREEN')
    log('')
    log('Usage:', 'YELLOW')
    log('  python start.py                    Interactive menu', 'WHITE')
    log('  python start.py <app>              Direct launch in debug mode', 'WHITE')
    log('  python start.py <app> debug        Direct launch in debug mode', 'WHITE')
    log('  python start.py <app> build        Direct launch in build mode', 'WHITE')
    log('  python start.py list               List available apps', 'WHITE')
    log('  python start.py help               Show this help', 'WHITE')
    log('')
    log('Available Apps:', 'YELLOW')
    for key in sorted(app_configs.keys()):
        config = app_configs[key]
        log(f'  - {key}', 'CYAN')
        log(f'    ({config["displayName"]} on port {config["port"]})', 'GRAY')
    log('')
    log('Examples:', 'YELLOW')
    log('  python start.py pymatrix           Start pyMatrix in debug mode', 'WHITE')
    log('  python start.py ittools build      Build ITTools', 'WHITE')
    log('')
    separator('GREEN')


# ============================================================
# Show List
# ============================================================
def show_list(app_configs: Dict):
    """Show list of available apps"""
    log('')
    separator('GREEN')
    log('  AVAILABLE APPLICATIONS', 'GREEN')
    separator('GREEN')
    log('')
    for key in sorted(app_configs.keys()):
        config = app_configs[key]
        log('  App Name     : ', 'YELLOW')
        log(f'  {key}', 'CYAN')
        log('  Display Name : ', 'YELLOW')
        log(f'  {config["displayName"]}', 'WHITE')
        log('  Port         : ', 'YELLOW')
        log(f'  {config["port"]}', 'WHITE')
        log('')
    separator('GREEN')


# ============================================================
# Execute Command with Error Handling
# ============================================================
def execute_command(command: list, description: str, cwd: Path, env: Dict) -> bool:
    """Execute command and return success status"""
    log('')
    separator('MAGENTA')
    log(f'  {description}', 'MAGENTA')
    separator('MAGENTA')
    log(f'[COMMAND] {" ".join(command)}', 'YELLOW')
    log('')

    try:
        # Merge environment
        command_env = os.environ.copy()
        command_env.update(env)

        subprocess.run(
            command,
            cwd=str(cwd),
            env=command_env,
            check=True,
            stdout=sys.stdout,
            stderr=sys.stderr
        )

        log('')
        log(f'[SUCCESS] {description} completed', 'GREEN')
        separator('MAGENTA')
        log('')
        return True

    except subprocess.CalledProcessError as e:
        log('')
        log(f'[ERROR] {description} failed', 'RED')
        log(f'[ERROR] Exit code: {e.returncode}', 'RED')
        separator('MAGENTA')
        log('')
        return False
    except Exception as e:
        log('')
        log(f'[ERROR] {description} failed: {e}', 'RED')
        separator('MAGENTA')
        log('')
        return False


# ============================================================
# Main Function
# ============================================================
def main():
    """Main entry point"""
    log('')
    separator('CYAN')
    log('  NUXT MAIN APPLICATION LAUNCHER - INITIALIZATION', 'CYAN')
    separator('CYAN')
    log('')
    log('[TRACE] Script Initialization:', 'YELLOW')
    log(f'  > Original Working Directory: {ORIGINAL_WORKING_DIR}', 'WHITE')
    log(f'  > Script Directory: {SCRIPT_DIR}', 'WHITE')
    log(f'  > Application Directory: {APP_DIR}', 'WHITE')
    log('')

    # Scan app configs
    log('[TRACE] Scanning application configurations...', 'YELLOW')
    app_configs = scan_app_configs()
    log(f'  [OK] Found {len(app_configs)} applications', 'GREEN')
    log('')

    # Change to app directory
    log('[TRACE] Changing Directory to Application Root:', 'YELLOW')
    log(f'  > Set Working Directory: {APP_DIR}', 'WHITE')
    os.chdir(APP_DIR)
    log(f'  [OK] Current Location: {os.getcwd()}', 'GREEN')
    log('')
    separator('CYAN')
    log('')

    # Parse arguments
    parsed_args = parse_args()

    if parsed_args['showHelp']:
        show_help(app_configs)
        sys.exit(0)

    if parsed_args['showList']:
        show_list(app_configs)
        sys.exit(0)

    # Determine mode
    selected_app = None
    mode = 'debug'
    selected_index = 0

    if parsed_args['appName']:
        # Command-line mode
        if parsed_args['appName'] not in app_configs:
            log('')
            separator('RED')
            log(f'  ERROR: Unknown application \'{parsed_args["appName"]}\'', 'RED')
            separator('RED')
            log('')
            log('Available apps:', 'YELLOW')
            for key in sorted(app_configs.keys()):
                log(f'  - {key}', 'CYAN')
            log('')
            sys.exit(1)

        selected_app = app_configs[parsed_args['appName']]
        mode = parsed_args['mode']

        log('')
        separator('GREEN')
        log('  COMMAND-LINE MODE ACTIVATED', 'GREEN')
        separator('GREEN')
        log(f'  Selected App: {selected_app["displayName"]}', 'CYAN')
        log(f'  Mode: {mode}', 'CYAN')
        separator('GREEN')
        log('')
    else:
        # Interactive mode
        saved_state = get_menu_state()
        menu_result = show_interactive_menu(app_configs, saved_state)
        selected_app = menu_result['selectedApp']
        mode = menu_result['mode']
        selected_index = menu_result['selectedIndex']
        save_menu_state(selected_index, mode)

    # Set environment variables
    env = {
        'NUXT_HOST': '0.0.0.0',
        'NUXT_PORT': str(selected_app['port']),
        'APP_ENTRY': selected_app['name']
    }

    log('')
    separator('CYAN')
    log('  APPLICATION STARTUP INFO', 'CYAN')
    separator('CYAN')
    log('')
    log('=== Application Selection ===', 'YELLOW')
    log(f'  Selected App     : {selected_app["displayName"]}', 'GREEN')
    log(f'  Namespace        : {selected_app["name"]}', 'GREEN')
    log(f'  Mode             : {mode}', 'GREEN')
    log(f'  Port             : {selected_app["port"]}', 'GREEN')
    log(f'  Host             : 0.0.0.0', 'GREEN')
    log('')
    log('=== Network Configuration ===', 'YELLOW')
    log(f'  Local URL        : http://127.0.0.1:{selected_app["port"]}', 'CYAN')
    log(f'  Network URL      : http://0.0.0.0:{selected_app["port"]}', 'CYAN')
    log('')
    separator('CYAN')
    log('')

    # Step 1: Switch pages directory
    switch_success = execute_command(
        ['node', str(SWITCH_APP_SCRIPT), selected_app['name']],
        'STEP 1: SWITCHING APP PAGES DIRECTORY',
        APP_DIR,
        env
    )

    if not switch_success:
        log('[ERROR] Failed to switch pages directory', 'RED')
        sys.exit(1)

    # Step 2: Start application server
    if mode == 'debug':
        log('')
        log('[INFO] Opening browser...', 'GREEN')
        url = f'http://127.0.0.1:{selected_app["port"]}'
        try:
            import webbrowser
            webbrowser.open(url)
        except Exception:
            log('[WARNING] Could not open browser automatically', 'YELLOW')

        execute_command(
            ['node', str(SWITCH_APP_SCRIPT), selected_app['name'], '--mode', 'dev'],
            'STEP 2: STARTING APPLICATION SERVER (DEBUG MODE)',
            APP_DIR,
            env
        )
    else:
        execute_command(
            ['node', str(SWITCH_APP_SCRIPT), selected_app['name'], '--mode', 'build'],
            'STEP 2: BUILDING APPLICATION (BUILD MODE)',
            APP_DIR,
            env
        )

    # Restore original directory
    os.chdir(ORIGINAL_WORKING_DIR)


# ============================================================
# Entry Point
# ============================================================
if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        log('\n[INFO] Interrupted by user', 'YELLOW')
        sys.exit(0)
    except Exception as e:
        log(f'\n[FATAL ERROR] {e}', 'RED')
        import traceback
        traceback.print_exc()
        sys.exit(1)
